add mape, smape, mad, spearman_r, spearmean_r_p_value

aaronspring opened the pull request:
add new metrics:
- mape
- smape
- mad
- spearman_r incl. p value

add new tests that should have been added before:
- [ ] test weights

todo:
- [x] test `np.deterministic`
- [x] test `deterministic`
- [x] check zeros in MAPE
- [x] how to handle number of nans in `union(a,b)` for `dof` in `*_r_p_value`
- [x] test `skipna`


addresses: https://github.com/raybellwaves/xskillscore/issues/37 https://github.com/raybellwaves/xskillscore/issues/32

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -112,6 +114,106 @@ def pearson_r(a, b, dim, weights=None):
+    Returns
+    -------
+    xr.DataArray or xr.Dataset
+        Pearson's correlation coefficient.
```
Spearman's

bradyrx replied:
Also change these returns to `xarray.DataArray or xarray.Dataset` for consistency with above. This is a typo from one of my docstrings that I fixed in #31 I believe.

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -112,6 +114,106 @@ def pearson_r(a, b, dim, weights=None):
+    Returns
+    -------
+    xr.DataArray or xr.Dataset
+        Pearson's correlation coefficient.
```
Spearman

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -286,3 +388,132 @@ def mae(a, b, dim, weights=None):
+
+def mad(a, b, dim, weights=None):
+    """
+    Mean Absolute Deviation.
```
Median

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -286,3 +388,132 @@ def mae(a, b, dim, weights=None):
+    --------
+    sklearn.metrics.median_absolute_error
+    xarray.apply_ufunc
+    https://en.wikipedia.org/wiki/Median_absolute_deviation
```
I agree this is useful but can you put it in a References section
https://github.com/scipy/scipy/blob/master/scipy/stats/stats.py#L930

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -286,3 +388,132 @@ def mae(a, b, dim, weights=None):
+    Returns
+    -------
+    xarray.Dataset or xarray.DataArray
+        Mean Absolute Error.
```
Symmetric Mean Absolute Percentage Error

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
+
+    See Also
+    --------
+    sklearn.metrics.mean_absolute_error / bo.absolute(a) * 100
```
what's `bo`.

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
+    See Also
+    --------
+    sklearn.metrics.mean_absolute_error / (a+b) * 100
+    https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error
```
Move to References please. Good to copy to def mape as well

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
+
+    See Also
+    --------
+    sklearn.metrics.mean_absolute_error / (a+b) * 100
```
a + b

raybellwaves commented on xskillscore/tests/test_deterministic.py:
```diff
@@ -343,3 +337,42 @@ def test_mae_xr_dask(a_dask, b_dask, dim, weight, weights_dask):
+    actual = spearman_r(a, b, dim)
+    # dirty fix, this only tests whether spearman_r is equal to pearson_r rankdata
+    # but this tests spearman_r doesnt crash on all AXES
+    if len(dim) == 1:
```
Should `test_spearman_r``` go in test_np_deterministic.py? We often have `test_METRIC_xr` and ```test_METRIC_xr_dask``` in `test_deterministic.py`

raybellwaves commented:
Looks good. Add tests for all the metrics in `test_np_deterministic.py`. Then add tests in `test_deterministic.py`.

bradyrx commented:
Also I imagine that #31 will go in before this, so make sure to rebase onto that and add the `skipna` keyword to the high-level functions here.

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -73,6 +75,69 @@ def _pearson_r(a, b, weights, axis):
+    return p
+
+
 def _pearson_r_p_value(a, b, weights, axis):
```
Should this be bumped back up to below `_pearson_r`?

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -73,6 +75,69 @@ def _pearson_r(a, b, weights, axis):
+    """
+    # https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764
+    rs = _spearman_r(a, b, weights, axis)
+    dof = a.shape[axis] - 2  # degrees of freedom
```
Maybe handle this as in the same way as `_pearson_r_p_value`:

```python
a = np.rollaxis(a, axis)
dof = a.shape[0] - 2
```

bradyrx replied:
Also I would either change this to `df` as in `_pearson_r_p_value` or change that one to `dof`. Probably the latter since `df` is like a pandas dataframe.

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
         return absolute_error.mean(axis=axis)
+
+
+def _mad(a, b, weights, axis):
```
Since this is the median (and there's no sum or mean funcs) this shouldn't take a `weights` argument.

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
+    sklearn.metrics.median_absolute_error
+
+    """
+    weights = _check_weights(weights)
```
As above, remove this line

bradyrx commented on xskillscore/core/deterministic.py:
```diff
@@ -286,3 +388,132 @@ def mae(a, b, dim, weights=None):
     )
+
+
+def mad(a, b, dim, weights=None):
```
Should drop the weights option. See comments in `np_deterministic`.

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -219,3 +284,107 @@ def _mae(a, b, weights, axis):
+
+    """
+    weights = _check_weights(weights)
+    # check whether a as zeros ?
```
Perhaps in high-level function with `xarray` funcs? How is this handled in sklearn?

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -87,6 +89,77 @@ def _pearson_r(a, b, weights, axis, skipna):
+    if skipna:
+        rankfunc = bn.rankdata
+    else:
+        rankfunc = bn.nanrankdata
```
Is this supposed to be under `if skipna`? Also curious to see if this works since the `bottleneck` documentation suggests this just returns nans as nans. Maybe you have to do `nanrankdata` and then some sort of `dropna` function before feeding it into `pearson_r`. You'll figure out in testing I guess

aaronspring commented:
I came across another issue in `np_deterministic.py` in `pearson_r` and `spearman_r` functions.
```
# TODO:  should dof be reduced by number of nans in a or b?
```
@bradyrx @raybellwaves

aaronspring commented on xskillscore/core/np_deterministic.py:
```diff
@@ -116,17 +130,94 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
+    # TODO:  should dof be reduced by number of nans in a or b?
+    dof = a.shape[0] - 2
+    t_squared = r ** 2 * (dof / ((1.0 - r) * (1.0 + r)))
+    _x = dof / (dof + t_squared)
```
I think this is problematic

aaronspring replied:
the dof only minus 2, the comment you made below

bradyrx commented:
> # TODO:  should dof be reduced by number of nans in a or b?

I think it should be the intersection of where `a` and `b` don't have NaNs. But you have to compute it at the right step so that you get the intersection, or compute in manually.

What I mean is the code should hopefully work such that in the following case

```python
A = [1, 2, nan, 4]
B = [1, nan, nan, 4]
```

The product of A and B would be reduced to: `[1, nan, nan, 16]`. So you'd have 2 dof there since the product of `nan` and `2` got nanned out. You should be able to count the length of the array after the product. Or do an intersect of some sort.

aaronspring commented:
Maybe we should rethink the way we implemented skipna for correlations. https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.spearmanr.html has an argument ˋnan_policyˋ describing how to deal with nans.

aaronspring commented:
fixed the tests now to my best knowledge. 
Changed `test_mask_skipna.py` and in particular `mask_data`.
Currently `spearman_r` gives a 1.0 correlation over nan-land. is this a problem?
ready for review

bradyrx commented:
>  Currently spearman_r gives a 1.0 correlation over nan-land. is this a problem?

`spearman` does this or the p-value? The pearson p value gives exactly 1.0 over land. You see that I anticipate this in the `skipna` testing here:

`assert (res_no_skipna == 1.0).all()`

I think the correlation itself should return nan but the p value should return 1.

aaronspring commented:
> > Currently spearman_r gives a 1.0 correlation over nan-land. is this a problem?
> 
> `spearman` does this or the p-value? The pearson p value gives exactly 1.0 over land. You see that I anticipate this in the `skipna` testing here:
> 
> `assert (res_no_skipna == 1.0).all()`
> 
> I think the correlation itself should return nan but the p value should return 1.

I think the same. But currently spearman_r (not the p value) gives 1 over land (see test_masked_skipna).

aaronspring commented:
I changed the masking in the tests also to better mimick land.

aaronspring commented:
I dont want to user spearman_r directly because it will be slower. rather I would like to rewrite it numpy fast. https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764

raybellwaves commented:
> I dont want to use scipy.stats.spearman_r directly because it will be slower. rather I would like to rewrite it numpy fast. [scipy/scipy:scipy/stats/stats.py@`v1.3.1`#L3613-L3764](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764) [](/scipy/scipy/blob/HEAD@{2019-10-30T21:57:46Z}/scipy/stats/stats.py#L3613-L3764)

I don't have any preference. It is noticeably slower? Mind creating a quick gist with times comparing both operations? Given this project is in it's infancy I would opt for code simplicity over speed.

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -116,17 +130,99 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
-    t_squared = r ** 2 * (df / ((1.0 - r) * (1.0 + r)))
-    _x = df / (df + t_squared)
+    b = np.rollaxis(b, axis)
+    # dirty fix: want to check the number of nans in union(a,b) on axis 0
```
Is this still valid? It seems counts a union of the nan's is a workable solution

aaronspring replied:
I am not sure whether this is really want it should be doing. And I am confused because I dont see this in scipy.stats.pearson_r.

aaronspring replied:
I should take away the #dirty fix comment. I think this implementation is ok

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -116,17 +130,99 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
+    b = np.rollaxis(b, axis)
+    # dirty fix: want to check the number of nans in union(a,b) on axis 0
+    nnans = np.sum(np.isnan(a[0] * b[0]))
+    dof = a.shape[0] - 2 - nnans
```
wonder if we can simply put an if statement `if np.allnan(a) ...` (or what ever it is) then return ...

aaronspring replied:
here I just want to reduce the dofs. another point raised elsewhere is whether if all datapoints are nan, the results for any correlation should be nan and for _p_value should be 1 or nan @bradyrx

raybellwaves replied:
Perhaps create an additional function called `_check_for_all_nans` which returns nan if a or b is full of nans. We can then use that in all the metrics. I'm not very familiar with warnings but we could raise a warning saying "array [a/b] is full of nans returning nan".

aaronspring commented:
> > I dont want to use scipy.stats.spearman_r directly because it will be slower. rather I would like to rewrite it numpy fast. [scipy/scipy:scipy/stats/stats.py@`v1.3.1`#L3613-L3764](https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764)
> 
> I don't have any preference. It is noticeably slower? Mind creating a quick gist with times comparing both operations? Given this project is in it's infancy I would opt for code simplicity over speed.

I decided to rather rewrite code with basic numpy commands as you did initially for `_pearson_r` instead of just wrapping `scipy.stats.pearsonr` with `xr.apply_ufunc` which I could have done here also for `scipy.stats.spearmanr`.

aaronspring commented:
I added an example demo notebook that shows the behaviour of metrics on `xr.tutorial.rasm`.
Current issues:
- `pearson_r_p_value` and `spearman_r_p_value` return nan everywhere
- `spearman_r` returns `1` over ocean (all nan over dim).

aaronspring commented:
now, finally, I see this PR ready. 
I changed `pearons_r_p_value` to get nans for masked input arrays. 
Without
```python
     all_nan = np.isnan(a.mean(axis=0) * b.mean(axis=0))
    # reset masked values to nan
    res = np.where(all_nan, np.nan, res)
```
the plots look ugly. Now all metrics give masked results for masked input.

There arae also tons of warnings in xskillscore. https://github.com/raybellwaves/xskillscore/issues/41

bradyrx commented:
I'll pull down https://github.com/raybellwaves/xskillscore/pull/38/commits/a75ec3f9f66c66bdc826824116bcae4aff2b0914, I was just looking at the notebook from the previous commit.

But this definitely looks like unexpected functionality (perhaps that's what you address):

<img width="635" alt="Screen Shot 2019-11-04 at 9 41 14 AM" src="https://user-images.githubusercontent.com/8881170/68139378-4466d080-fee7-11e9-93d1-bd22bd99a5ff.png">

We shouldn't get that rectangle with skipna. The main point of skipna was for dealing with applications over space dimensions where space is compressed into a single time series-like object and the nans mess it up.

<img width="647" alt="Screen Shot 2019-11-04 at 9 42 21 AM" src="https://user-images.githubusercontent.com/8881170/68139513-77a95f80-fee7-11e9-936f-17ac1825371d.png">

bradyrx commented:
Yeah even with the new commit `skipna=True` creates that box there. But now that I think of it, this is over the time dimension so it's just ignoring the first time step because it's the only one with nans, correct? I think this is the proper functionality.

aaronspring commented:
> We shouldn't get that rectangle with skipna. The main point of skipna was for dealing with applications over space dimensions where space is compressed into a single time series-like object and the nans mess it up.

I think the box should there.

aaronspring commented:
yes. skipna=True ignores this first timestep, while skipna=False will return nan.

bradyrx commented:
Also `docs/source/examples/.ipynb_checkpoints/Demo-checkpoint.ipynb` was added to this and should not be.

bradyrx commented:
Okay I'm good after those two points are addressed. Of course I think the notebook needs a lot more descriptive text, etc., but this is a good first stake in the ground and that can be addressed on a docs PR.

@raybellwaves I think you would benefit from having `reviewnb` installed on this repo so we can comment directly on notebooks that are in PRs. I use it on the other repos and it works well -- you don't have to clone the repo and run the notebook yourself.

raybellwaves commented:
Thanks @aaronspring. I'll take a look this weekend

raybellwaves commented:
@aaronspring mind moving the notebook out from `docs/source/examples/Demo.ipynb` and putting it in `examples/Demo.ipynb`. We'll leave the `docs` folder alone until we setup the documentation (https://github.com/raybellwaves/xskillscore/issues/36)

raybellwaves commented on ci/requirements-py36.yml:
```diff
@@ -12,6 +12,9 @@ dependencies:
   - bottleneck
+  - matplotlib
+  - cftime==1.0.3.4
+  - nc-time-axis
```
Mind removing these with the notebook? or push the commits to another branch for another PR.

aaronspring replied:
Ok

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -291,7 +414,137 @@ def mae(a, b, dim, weights=None, skipna=False):
+    See Also
+    --------
+    xarray.apply_ufunc
+
```
Mind copying the Reference from `_smape` e.g.

Reference
    ---------
    https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -291,7 +414,137 @@ def mae(a, b, dim, weights=None, skipna=False):
+    See Also
+    --------
+    xarray.apply_ufunc
+
```
Mind copying the Reference from _mmape e.g.

    Reference
    ---------
    https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

raybellwaves commented on xskillscore/core/deterministic.py:
```diff
@@ -159,8 +180,110 @@ def pearson_r_p_value(a, b, dim, weights=None, skipna=False):
+    --------
+    scipy.stats.spearman_r
+    xarray.apply_ufunc
+    """
```
Mind copying the Reference from _spearman_r e.g.

    Reference
    ---------
    https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764

raybellwaves commented on xskillscore/core/np_deterministic.py:
```diff
@@ -116,17 +130,96 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
     res = special.betainc(_a, _b, _x)
+    # reset masked values to nan
+    all_nan = np.isnan(a.mean(axis=0) * b.mean(axis=0))
+    res = np.where(all_nan, np.nan, res)
```
Thanks for the fix
I wanted to make sure it was doing what I thought it should
Could add this to testing
https://gist.github.com/raybellwaves/3834b1005094dd748d89aaf6875ca531

aaronspring replied:
This is already included in test_masked_metric

raybellwaves commented on xskillscore/tests/test_deterministic.py:
```diff
@@ -116,19 +118,26 @@ def adjust_weights(dim, weight, weights):
-def test_pearson_r_xr(a, b, dim, weight, weights):
+@pytest.mark.parametrize('metrics', correlation_metrics)
+@pytest.mark.parametrize('dim', AXES)
+@pytest.mark.parametrize('weight', [True, False])
```
mind renaming `weight` here to `weight_bool`.

raybellwaves commented on xskillscore/tests/test_mask_skipna.py:
```diff
@@ -2,104 +2,89 @@
+@pytest.mark.parametrize('dim', AXES)
+def test_metrics_masked(a, b, dim, metric):
+    """Test for all distance-based metrics whether result if skipna do not
+     contain any nans when applied along dim with nans."""
```
some typos """Test for all distance-based metrics whether result of skipna does not
     contain any nans when applied along dim with nans."""

aaronspring commented:
I will address the requested changes tomorrow.

aaronspring commented:
PR ready to merge

raybellwaves commented:
Thanks @aaronspring !
